### PR TITLE
Feature/sg 1351 similarity feature

### PR DESCRIPTION
**Pull request by shaydeci**

### CODE REVIEW

File: src/data_gradients/feature_extractors/object_detection/similarity.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
+from typing import List, Optional
+
+import numpy as np
+from torch.nn import Identity
+from torchvision.models import swin_v2_b, Swin_V2_B_Weights
```
Comment (BloodAxe) [inline src/data_gradients/feature_extractors/object_detection/similarity.py:5]: I think we should add `torchvision` to requirements.txt with version specifier `>=` corresponding to a version where Swin was introduced.

**Review by ofrimasad — COMMENTED** [review]
minor comment. looks very nice.

File: src/data_gradients/feature_extractors/object_detection/similarity.py (hunk 1 above)
Comment (shaydeci) [reply src/data_gradients/feature_extractors/object_detection/similarity.py:5]: 100 percent, totally forgot.
Thanks!

Comment (shaydeci) [reply src/data_gradients/feature_extractors/object_detection/similarity.py:5]: b64828c3ce5e111def6aa32e59a76024cd3a6d62

File: src/data_gradients/feature_extractors/object_detection/similarity.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        """
+        Extracts feature vectors from a list of cropped images using the pre-trained model.
+
+        :param cropped_images: List[np.ndarray]. List of cropped images from detected objects.
+        :param image_channels: ImageChannels. When BGRChannels, will reverse the channels.
+        :return: List[torch.Tensor]. Extracted feature vectors for each image.
+        """
+        # Process each image individually since crops are of varying sizes
+        if isinstance(image_channels, BGRChannels):
+            cropped_images = [c[:, :, ::-1].copy() for c in cropped_images]
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/data_gradients/feature_extractors/object_detection/similarity.py:57]: I raise in error if the format is not BGR or RGB, as this is what the model expects.
I can add support for grayscaleby truncating the channels.

File: src/data_gradients/feature_extractors/object_detection/similarity.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        image_height, image_width = image.shape[:2]
+        cropped_images = []
+        class_ids = []
+        iou_matrix = None
+        image_channels = sample.image_channels
+        if not isinstance(image_channels, BGRChannels) and not isinstance(image_channels, RGBChannels):
+            raise RuntimeError(f"Similarity feature only works with RGB and BGR samples, got {image_channels}")
+
+        if self.all_classes_list is None:
+            self.all_classes_list = sample.class_names
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/data_gradients/feature_extractors/object_detection/similarity.py:82]: entire dataset

File: src/data_gradients/feature_extractors/object_detection/similarity.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        """
+        # Process each image individually since crops are of varying sizes
+        if isinstance(image_channels, BGRChannels):
+            cropped_images = [c[:, :, ::-1].copy() for c in cropped_images]
+        cropped_images = [torch.from_numpy(c.transpose((2, 0, 1))).type(torch.FloatTensor) for c in cropped_images]
+        cropped_images = [c / 255 for c in cropped_images]
+        cropped_images = [self.image_preprocessor(c) for c in cropped_images]
+        cropped_images = torch.stack(cropped_images)
+        with torch.no_grad():
+            extracted_features = self.model(cropped_images)
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/data_gradients/feature_extractors/object_detection/similarity.py:63]: I tested passing 200 and didn't have any issue locally.
Notice this runs on cpu, COOM is not an issue here.

**Review by BloodAxe — APPROVED** [review]
LGTM

File: src/data_gradients/feature_extractors/object_detection/similarity.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        :param sample: DetectionSample. The new sample to process, containing image data and bounding boxes.
+        """
+        image = sample.image  # np.ndarray of shape [H, W, C] - The image as a numpy array with channels last
+        image_height, image_width = image.shape[:2]
+        cropped_images = []
+        class_ids = []
+        iou_matrix = None
+        image_channels = sample.image_channels
+        if not isinstance(image_channels, BGRChannels) and not isinstance(image_channels, RGBChannels):
+            raise RuntimeError(f"Similarity feature only works with RGB and BGR samples, got {image_channels}")
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/data_gradients/feature_extractors/object_detection/similarity.py]: 648409ad457f5a1f7eebfbc57d8b2db7b4a23b22 added

File: src/data_gradients/feature_extractors/object_detection/similarity.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/data_gradients/feature_extractors/object_detection/similarity.py:57]: 648409ad457f5a1f7eebfbc57d8b2db7b4a23b22

**Pull request merged**